### PR TITLE
Restructure Node class

### PR DESCRIPTION
**Pull request by mojoe12**
restructuring Node class

### CODE REVIEW

File: src/TreePiece.h [hunk 1]
```diff
@@ -487,7 +487,7 @@ void TreePiece<Data>::perturb (Real timestep, bool if_flush) {
       Vector3D<Real> old_position = particle.position;
       particle.perturb(timestep, leaf->getForce(i), readers.ckLocalBranch()->universe.box);
       //CkPrintf("magitude of displacement = %lf\n", (old_position - leaf->particles[i].position).length());
-      //CkPrintf("total centroid is (%lf, %lf, %lf)\n", global_root->data.getCentroid().x, global_root->data.getCentroid().y, global_root->data.getCentroid().z);
+      //CkPrintf("total centroid is (%lf, %lf, %lf)\n", global_root->data.centroid.x, global_root->data.centroid.y, global_root->data.centroid.z);
```
Comment (Wingpad) [inline src/TreePiece.h:490]: Can we simply remove these debug print statements or put them under some kind of verbosity control? Perhaps an #ifdef?

Comment (Wingpad) [reply src/TreePiece.h:490]: More of a general question than anything directly pertaining to this commit.

File: src/CacheManager.h [hunk 2]
```diff
@@ -84,26 +84,22 @@ void CacheManager<Data>::startPrefetch(DPHolder<Data> dp_holder, CkCallback cb)
...
   Key curr_key = node->key;
-  while (curr_key > 1) {
-    curr_key /= BRANCH_FACTOR;
+  auto branch_factor = node->getBranchFactor();
+  while (curr_key >= 1) {
+    curr_key /= branch_factor;
     prefetch_set.insert(curr_key);
+    for (int i = 0; i < branch_factor; i++) {
+      prefetch_set.insert(curr_key * branch_factor + i);
+    }
```
Comment (minitu) [inline src/CacheManager.h:102]: Wouldn't this be problematic when `curr_key` is 1 (the root)? After `curr_key /= branch_factor`, `curr_key` would be 0.

Comment (mojoe12) [reply src/CacheManager.h:102]: yeah can you fix it

File: src/CountVisitor.h [hunk 3]
```diff
@@ -30,23 +30,23 @@ class CountVisitor {
   }
 
 public:
-  bool node(SourceNode<CentroidData> from, TargetNode<CentroidData> on) {
-    if (from.data.count == 0 || on.data->count == 0) {
+  bool node(const SpatialNode<CentroidData>& from, SpatialNode<CentroidData>& on) {
```
Comment (minitu) [inline src/CountVisitor.h:33]: Should `on` also have the const qualifier?

Comment (mojoe12) [reply src/CountVisitor.h:33]: on needs to be edited. for example, in GravityVisitor we call target.applyForce

**Pull request merged**